Include/internal/pycore_dynarray.h

```diff
@@ -0,0 +1,143 @@
+#ifndef Py_INTERNAL_DYNARRAY_H
+#define Py_INTERNAL_DYNARRAY_H
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "Python.h" // Py_ssize_t
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#define _PyDynArray_DEFAULT_SIZE 16
+
+typedef void (*_PyDynArray_Deallocator)(void *);
+
+typedef struct {
+    void **items;
+    Py_ssize_t capacity;
+    Py_ssize_t length;
+    _PyDynArray_Deallocator deallocator;
```

Comment:
How about taking an idea from `Py_hastable_t'?

It has a `Py_hashtable_allocator_t alloc' field which can be used both for allocation and deallocation.

Reply:
I decided that's too inconvenient for users. In basically all cases, there's no need to customize the allocator. We're using the `Raw` domain anyway, so it's not like we have to worry about the caller holding the GIL.